File: javadarts/src/main/java/com/sdoward/barstewarddarts/android/java/TurnValidator.java
package com.sdoward.barstewarddarts.android.java;

public class TurnValidator {

    private Throw firstThrow = null;
    private Throw secondThrow = null;
    private Throw thirdThrow = null;

    public void acceptThrow(Throw frow) {
        if (firstThrow == null) {
            firstThrow = frow;
        } else if (secondThrow == null) {
            secondThrow = frow;
        } else if (thirdThrow == null) {
            thirdThrow = frow;
        } else {
            throw new IllegalStateException("3 throws are complete. ");
        }
    }

    public boolean turnIsComplete() {
        return firstThrow != null && secondThrow != null && thirdThrow != null;
    }


    public Turn getTurn(){
        if (turnIsComplete()) {
            Turn turn = Turn.create(firstThrow, secondThrow, thirdThrow);
            resetTurn();
            return turn;
        } else {
            throw new IllegalStateException("3 throws haven't been completed");
        }

    }

    private void resetTurn() {
        firstThrow = null;
        secondThrow = null;
        thirdThrow = null;
    }

}

File: kotlindarts/src/main/kotlin/com/sdoward/barstewarddarts/android/kotlin/TurnValidator.kt
package com.sdoward.barstewarddarts.android.kotlin

object TurnValidator {

    private var firstThrow: Throw? = null
    private var secondThrow: Throw? = null
    private var thirdThrow: Throw? = null

    fun acceptThrow(frow: Throw) {
        if (firstThrow == null) {
            firstThrow = frow
        } else if (secondThrow == null) {
            secondThrow = frow
        } else if (thirdThrow == null) {
            thirdThrow = frow
        } else {
            throw IllegalStateException("3 throws are complete. ")
        }

    }

    fun turnIsComplete() :Boolean {
        return firstThrow !== null && secondThrow !== null && thirdThrow !== null
    }

    val turn: Turn
        get() {
            if (turnIsComplete()) {
                val turn = Turn(firstThrow!!, secondThrow!!, thirdThrow!!)
                resetTurn()
                return turn
            } else {
                throw java.lang.IllegalStateException("3 throws haven't been completed")
            }

        }
     fun resetTurn() {
        firstThrow = null
        secondThrow = null
        thirdThrow = null
    }


}